### PR TITLE
Add very basic skin editor undo / redo support

### LINKED ISSUE #12966: Add ability to undo in skin editor
**Describe the new feature:** Add ability to undo change (move, scale, rotate etc.) with a shortcut like CTRL+Z (or Cmd+Z in Mac).

### PR DESCRIPTION
**Pull request by peppy**
- Closes https://github.com/ppy/osu/issues/12966

I made this work and it seems like it might be okay to get it in in this state? I'm hesitant to write tests or spend more time on this due to the fact that https://github.com/ppy/osu/issues/22388 will likely require deep modifications in how change handling is done (assuming we move every target container to a single json structure or similar).

Put another way, I wasn't sure if I should work on undo/redo support before considering the aforementioned restructure, but getting it working to a MVP level was quick enough that I propose we get this out to users so they can benefit from it while the larger restructuring tasks happen in the background.

Of note: if any new elements are added or removed, *all elements will be recycled*. This is the simplest way of handling this, especially since we don't have any ID for tracking movement of elements around. There may be room to use `SkinnableTargetContainer`'s [bindable `Components` list](https://github.com/ppy/osu/blob/2f30306ea204dfe8809ff0cb26effe6284a0d41a/osu.Game/Skinning/SkinnableTargetContainer.cs#L20) to better handle this in the future.

https://user-images.githubusercontent.com/191335/216571607-ad9e1a0e-47ce-4fff-a06b-ef1cbff1e205.mp4

PRing this to get some initial feedback on direction.

Issues of note:

- [x] Changes to `SettingsSourceAttribute`s do not trigger state saves.

### CODE REVIEW

File: osu.Game/Screens/Edit/EditorChangeHandler.cs [hunk 1]
```diff
@@ -1,31 +1,25 @@
...
-using osu.Game.Beatmaps.Formats;
-using osu.Game.Rulesets.Objects;
 
 namespace osu.Game.Screens.Edit
 {
     /// <summary>
     /// Tracks changes to the <see cref="Editor"/>.
     /// </summary>
-    public partial class EditorChangeHandler : TransactionalCommitComponent, IEditorChangeHandler
+    public abstract partial class EditorChangeHandler : TransactionalCommitComponent, IEditorChangeHandler
```
Comment (bdach) [inline osu.Game/Screens/Edit/EditorChangeHandler.cs:16]: With the way this is being abstracted away further, I'm not sure that the namespace or the xmldoc here make much sense.

Namespace comment also applies to `IEditorChangeHandler` and `TransactionalCommitComponent`.

File: osu.Game/Screens/Edit/EditorChangeHandler.cs [hunk 2]
```diff
@@ -122,6 +96,10 @@ public void RestoreState(int direction)
             updateBindables();
         }
 
+        protected abstract void WriteCurrentStateToStream(MemoryStream stream);
+
+        protected abstract void ApplyStateChange(byte[] previousState, byte[] newState);
```
Comment (bdach) [inline osu.Game/Screens/Edit/EditorChangeHandler.cs:101]: These need xmldoc.

File: osu.Game/Skinning/ISkinnableTarget.cs [hunk 3]
```diff
@@ -36,6 +36,11 @@ public interface ISkinnableTarget : IDrawable
         /// </summary>
         void Reload();
 
+        /// <summary>
+        /// Reload this target from the provided skinnable information.
+        /// </summary>
+        public void Reload(SkinnableInfo[] skinnableInfo);
```
Comment (bdach) [inline osu.Game/Skinning/ISkinnableTarget.cs:42]:
```suggestion
        void Reload(SkinnableInfo[] skinnableInfo);
```

**Review by bdach — COMMENTED** [review]
Generally code seems inoffensive on a read-through, some structure issues from me though

File: osu.Game/Screens/Edit/EditorChangeHandler.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Screens/Edit/EditorChangeHandler.cs:16]: Honestly wasn't sure if we should still keep the underlying `TransactionalCommitComponent`, but left it there to reduce changes for the time being. I'd be okay with moving this to another namespace, but I'm honestly not sure where they should exist. Open to suggestion.

Comment (bdach) [reply osu.Game/Screens/Edit/EditorChangeHandler.cs:16]: Let's leave this where it is until the dust settles and we see the final shape of the skin editor better for now I guess.

**Comment by bdach** [comment]
I can break this pretty hard using the "scene library" toolbar.

1. Start at main menu
2. Click the "Gameplay" scene from the library
3. Wait for it to load
4. Make any change
5. Notice that you can undo twice, and the second undo erases every component from the skin

https://user-images.githubusercontent.com/20418176/217051497-61e2052f-e133-427f-b93e-36ef09800d2a.mp4

It appears that the initial save in the skin editor change handler fails when doing this (resulting json is `{}`), because the `firstTarget` is null.

Careful not to trip over https://github.com/ppy/osu/issues/22263 and https://github.com/ppy/osu/issues/17391 when reproducing.

I also got this to crash once on a mismatched  `EndChange()` without `StartChange()` call, but I'm not completely sure what I did to make it happen, so I'm not sure I'll be able to make it happen again.

**Comment by bdach** [comment]
Never mind, I repro'd the second thing immediately.

1. Go to gameplay
2. Seek track to be close to end
3. Open skin editor
4. Select a component and start dragging it with left mouse
5. Wait for map to end, progressing to results screen
6. Release mouse to :boom: 

https://user-images.githubusercontent.com/20418176/217053059-822f6e17-721e-45f7-8a0a-a727cb2712c7.mp4

Crash trace is, unsurprisingly:

```
Unhandled exception. System.InvalidOperationException: Cannot call EndChange without a previous call to BeginChange.
   at osu.Game.Screens.Edit.TransactionalCommitComponent.EndChange() in /home/dachb/Documents/opensource/osu/osu.Game/Screens/Edit/TransactionalCommitComponent.cs:line 49
   at osu.Game.Overlays.SkinEditor.SkinEditor.EndChange() in /home/dachb/Documents/opensource/osu/osu.Game/Overlays/SkinEditor/SkinEditor.cs:line 485
   at osu.Game.Screens.Edit.Compose.Components.BlueprintContainer`1.OnDragEnd(DragEndEvent e) in /home/dachb/Documents/opensource/osu/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs:line 212
   at osu.Framework.Graphics.Drawable.TriggerEvent(UIEvent e)
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
```

**Review by bdach — CHANGES_REQUESTED** [review]
see above

**Comment by peppy** [comment]
> It appears that the initial save in the skin editor change handler fails when doing this (resulting json is `{}`), because the `firstTarget` is null.

This actually goes a bit deeper. `firstTarget` is non-null but the components haven't finished loading at the point the state is saved. `SkinBlueprintContainer` gets around this by watching for changes in components, but it has an easier task because it doesn't need to know which change is the *initial* one.. so if we used that method it would be best-effort..

https://github.com/ppy/osu/blob/26efb8e8423ccc7e4aa559d2aea28090d4fc3b61/osu.Game/Overlays/SkinEditor/SkinBlueprintContainer.cs#L56-L62

**Comment by peppy** [comment]
I've fixed both the issues you found. See what you think of the solutions.

**Comment by bdach** [comment]
Changes look okay (the `ensureStateSaved()` thing feels a little wonky but looks correct), issues look fixed, but e162fd56da39f5bcf67ef3a21af13ea08c6cf8ec broke tests and 0320ba770f1824b32fbb1e7d57366e7cf4b41481 broke code quality.

On a quick look the first one may not be very easy to fix, may just need to update assertions? Second issue should be trivial to resolve (not sure where that conditional access in the ctor there even comes from).

**Comment by bdach** [comment]
Seems solid enough for an initial implementation as far as I'm concerned.

I did find one issue with the "Revert to default" option and how it interacts with undo, but I think it's relatively minor and don't really want to be blocking this PR too much longer, so I'll open a follow-up issue for it.

**Review by bdach — APPROVED** [review]

**Comment by bdach** [comment]
Tracking aforementioned issue as https://github.com/ppy/osu/issues/22562

**Pull request merged**